File: leetcode_solution/3_Longest_Substring_Without_Repeating_Characters.java
import java.util.Arrays;
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int length = s.length();
        if(length == 0){
            return 0;
        }else if(length == 1){
            return 1;
        }else{

            int[] record = new int[length];
            Arrays.fill(record, 1);
            int i = 0;
            int j = 0;
            while(i <= length){
                while(i <= length - 1 && !(s.substring(j,i).contains(Character.toString(s.charAt(i))))){
                    i ++;
                }
                record[j] = i - j;
                if(i == length){
                    break;
                }else{
                    j = i - 1;
                    while(!s.charAt(i).equals(s.charAt(j))){
                        j --;
                    }
                    j++;
                }
            }
            int max = Arrays.stream(record).max().getAsInt();
            return max;

        }


    }
}

File: leetcode_solution/3_Longest_Substring_Without_Repeating_Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        length = len(s);
        if length == 0:
            return 0
        elif length == 1:
            return 1
        else:
            record = [1]*length;
           
            i = 0
            j = 0
            while(i <= length ):
                while(i<= length -1  and s[i] not in s[j:i]):
                    i += 1;
                record[j] = i - j
                if(i == length):
                    break;
                else:
                    j = i-1

                    # greedy
                    while(s[i] != s[j]):
                        j -= 1
                    j += 1

            return max(record)
